### DISCUSSION TITLE
`mechanize` + `ema` usage

### DISCUSSION THREAD

**Discussion by aaarrti** [google-deepmind/optax#585] [Q&A] [answered]
Hi, 

I want to use `optax.contrib.mechanize` together with the exponential moving average. 
What would be the correct way to do it? 
Option 1:
```python
tx = optax.contrib.mechanize(optax.chain(optax.ema(...), optax.lion(...)))
```
Option 2:
```python
tx = optax.chain(optax.ema(...), optax.contrib.mechanize(optax.lion(...)))
```

I would also appreciate it much, if you could share a bit of theory/background on why an option is correct, or just share any related resources. Thanks in advance!

**Comment by fabianp** [comment]
Perhaps @adefazio has some insights on this one?

**Comment by adefazio** [comment, marked as answer]
Mechanic is designed to work with any optimization algorithm as the base optimizer, so you should use the EMA within Mechanic (option 1) rather than wrapping it. The LR estimation may not work correctly if you apply the EMA via option 2, or at least the theory doesn't tell us what will happen.
